File: static/js/text-stats.js
/**
 * AuthentiText: quick document statistics (browser + Node).
 *
 * JavaScript twin of analyzer/services/text_stats.py. Both are tested against
 * analyzer/tests/fixtures/text_stats_cases.json, so the live counts in the
 * editor match what the server saves. See the Python module for the rules.
 * These are estimates; Phase 4 segments sentences with spaCy.
 */
(function (root, factory) {
  const api = factory();
  if (typeof module === "object" && module.exports) module.exports = api;  // Node (tests)
  else root.TextStats = api;                                             // Browser
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  const READING_WPM = 238;
  const WORD_RE = /[0-9]+(?:[.,][0-9]+)+|[\p{L}\p{N}]+(?:['\u2019-][\p{L}\p{N}]+)*/gu;
  const HAS_WORD_RE = /[\p{L}\p{N}]/u;
  const BLANK_LINE_RE = /\n[ \t]*\n/;
  const PARAGRAPH_SPLIT_BLANK = /\n\s*\n/;
  // Soft wrap: a line break inside a sentence, as when text is pasted from a PDF.
  // Mirrors analyzer/services/segmentation.py.
  const SENTENCE_END_RE = /[.!?\u2026:;]["'\u201d\u2019)\]]*$/;
  const LIST_MARKER_RE = /^\s*(?:[-*\u2022\u2013\u2014]|\d+[.)]|[a-z][.)]\s|#{1,6}\s)/;
  const CONTINUES_RE = /^["'\u201c\u2018(\[]*[a-z0-9]/;

  function isSoftWrap(before, after) {
    before = before.replace(/\s+$/, "");
    after = after.trim();
    if (!before || !after) return false;
    if (SENTENCE_END_RE.test(before) || LIST_MARKER_RE.test(after)) return false;
    return CONTINUES_RE.test(after);
  }
  const SENTENCE_SPLIT_RE = /(?<=[.!?\u2026])["'\u201d\u2019)\]]*\s+/u;

  function normalizeNewlines(text) {
    return text.replace(/\r\n?/g, "\n");
  }

  function splitParagraphs(text) {
    let blocks;
    if (BLANK_LINE_RE.test(text)) {
      blocks = text.split(PARAGRAPH_SPLIT_BLANK);
    } else {
      blocks = [];
      let current = "";
      text.split("\n").forEach((line, index) => {
        if (index && !isSoftWrap(current, line)) {
          blocks.push(current);
          current = line;
        } else {
          current = index ? `${current}\n${line}` : line;
        }
      });
      blocks.push(current);
    }
    return blocks.filter((block) => HAS_WORD_RE.test(block));
  }

  function splitSentences(paragraph) {
    return paragraph.split(SENTENCE_SPLIT_RE).filter((piece) => HAS_WORD_RE.test(piece));
  }

  function compute(text) {
    text = normalizeNewlines(text || "");
    const words = (text.match(WORD_RE) || []).length;
    const paragraphs = splitParagraphs(text);
    const sentences = paragraphs.reduce((sum, p) => sum + splitSentences(p).length, 0);
    return {
      characters: [...text].length, // code points, matching Python's len()
      words,
      sentences,
      paragraphs: paragraphs.length,
      reading_minutes: words ? Math.ceil(words / READING_WPM) : 0,
    };
  }

  return { compute, READING_WPM };
});

File: analyzer/services/text_stats.py
"""
Quick document statistics for the editor.

These are fast, rule-based ESTIMATES. They power the live counts on the Analyze
page and the counts saved with a document before the NLP pipeline runs.
Phase 4 replaces sentence and paragraph segmentation with spaCy, which
handles abbreviations ("e.g.", "Dr.") that these rules can't.

The same rules are implemented in static/js/text-stats.js. Both are tested
against analyzer/tests/fixtures/text_stats_cases.json so they can't drift apart.

Rules
- word:       letters/digits, allowing internal apostrophes or hyphens (it's, well-known);
              numbers with . or , separators count once (3.14, 1,000)
- paragraph:  if the text contains a blank line, blank lines separate paragraphs;
              otherwise each line break does. Only blocks containing a word count.
- sentence:   within each paragraph, a split after . ! ? or ... (optionally followed
              by closing quotes/brackets) and whitespace. Only pieces with a word count.
- characters: Unicode code points, with CRLF counted as one line break
- reading:    whole minutes, rounded up, at READING_WPM
"""
from __future__ import annotations

import math
import re
from dataclasses import asdict, dataclass

READING_WPM = 238  # average silent reading speed for non-fiction (Brysbaert, 2019)

# Numbers with separators (3.14, 1,000) first, then words with internal ' or -.
WORD_RE = re.compile(r"[0-9]+(?:[.,][0-9]+)+|[^\W_]+(?:['\u2019\-][^\W_]+)*")
HAS_WORD_RE = re.compile(r"[^\W_]")
BLANK_LINE_RE = re.compile(r"\n[ \t]*\n")
PARAGRAPH_SPLIT_BLANK = re.compile(r"\n\s*\n")
SENTENCE_SPLIT_RE = re.compile(r"(?<=[.!?\u2026])[\"'\u201d\u2019)\]]*\s+")


@dataclass(frozen=True)
class TextStats:
    characters: int
    words: int
    sentences: int
    paragraphs: int
    reading_minutes: int

    def as_dict(self) -> dict:
        return asdict(self)


def normalize_newlines(text: str) -> str:
    """Browsers submit textareas with CRLF; count them as a single line break."""
    return text.replace("\r\n", "\n").replace("\r", "\n")


def split_paragraphs(text: str) -> list[str]:
    splitter = PARAGRAPH_SPLIT_BLANK if BLANK_LINE_RE.search(text) else re.compile(r"\n")
    return [block for block in splitter.split(text) if HAS_WORD_RE.search(block)]


def split_sentences(paragraph: str) -> list[str]:
    return [piece for piece in SENTENCE_SPLIT_RE.split(paragraph) if HAS_WORD_RE.search(piece)]


def compute_text_stats(text: str) -> TextStats:
    text = normalize_newlines(text)
    words = len(WORD_RE.findall(text))
    paragraphs = split_paragraphs(text)
    sentences = sum(len(split_sentences(p)) for p in paragraphs)
    return TextStats(
        characters=len(text),
        words=words,
        sentences=sentences,
        paragraphs=len(paragraphs),
        reading_minutes=math.ceil(words / READING_WPM) if words else 0,
    )
